src/components/TodoItem.tsx

```diff
@@ -0,0 +1,262 @@
+import {
+  useContext, useEffect, useRef, useState,
+} from 'react';
+import cn from 'classnames';
+
+import { Todo } from '../types/Todo';
+import { DispatchContext, StateContext } from './TodosProvider';
+import { deleteTodo, updateTodo } from '../api/todos';
+import { ErrorMessage } from '../types/ErrorMessage';
+import { LoadingStatus } from '../types/LoadingStatus';
+import { TEMP_TODO_ID } from '../constants/tempTodoId';
+
+type Props = {
+  todo: Todo,
+};
+
+export const TodoItem:React.FC<Props> = ({ todo }) => {
+  const {
+    completed,
+    id,
+    title,
+    userId,
+  } = todo;
+
+  const {
+    todos,
+    shouldLoading,
+    newTodoInputRef,
+  } = useContext(StateContext);
+  const dispatch = useContext(DispatchContext);
+  const [isLoading, setIsLoading] = useState(!todos.find(t => t.id === id));
+  const [isUpdatingTitle, setIsUpdatingTitle] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+  const updateTitleRef = useRef<HTMLInputElement | null>(null);
+
+  const toggleChecked = async (e: React.ChangeEvent<HTMLInputElement>) => {
+    try {
+      setIsLoading(true);
+      const updatedTodo = await updateTodo({
+        id,
+        title,
+        completed: e.target.checked,
+        userId,
+      });
+
+      dispatch({
+        type: 'updateTodo',
+        payload: { todo: updatedTodo },
+      });
+    } catch (error) {
+      dispatch({
+        type: 'error',
+        payload: { error: ErrorMessage.Updating },
+      });
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const deleteCurrentTodo = async () => {
+    try {
+      setIsLoading(true);
+      await deleteTodo(id);
+
+      dispatch({
+        type: 'deleteTodo',
+        payload: { id },
+      });
+    } catch (error) {
+      dispatch({
+        type: 'error',
+        payload: { error: ErrorMessage.Deleting },
+      });
+      setNewTitle(title);
```

Comment:
as a question - why do you need to set a new title on error?

Reply:
Oh, it's probable just a typo)